File: backend/app/services/chat_service.py
"""
Chat service — orchestrates messaging, channels, read receipts, mentions.

Sits between the router and the 5 chat repos, adding:
  - Membership validation before send/read
  - Auto-enrollment of job crew into job channels
  - Notification dispatch for messages and @mentions
  - Inbox assembly with unread counts + last message preview
"""

from __future__ import annotations

import logging

import aiosqlite

from app.repositories.chat_repo import (
    ChatChannelRepo,
    ChatMemberRepo,
    ChatMessageRepo,
    ChatMentionRepo,
    ChatReadReceiptRepo,
)
from app.services.notification_service import NotificationService

logger = logging.getLogger(__name__)


class ChatService:
    """Orchestrates chat operations with business logic."""

    def __init__(self, db: aiosqlite.Connection) -> None:
        self.db = db
        self.channel_repo = ChatChannelRepo(db)
        self.member_repo = ChatMemberRepo(db)
        self.message_repo = ChatMessageRepo(db)
        self.read_repo = ChatReadReceiptRepo(db)
        self.mention_repo = ChatMentionRepo(db)
        self.notif_service = NotificationService(db)

    # ── Channels ───────────────────────────────────────────────────

    async def get_or_create_job_channel(
        self, job_id: int, user_id: int
    ) -> dict:
        """Get or create a job channel and auto-enroll crew members.

        When a job channel is first created, we pull all workers from
        job_team_members and job_dispatch to auto-populate the member list.
        """
        channel = await self.channel_repo.get_or_create_job_channel(
            job_id, user_id
        )

        # Auto-add crew members if they aren't already in the channel
        crew_ids = await self._get_job_crew(job_id)
        for crew_id in crew_ids:
            if not await self.member_repo.is_member(channel["id"], crew_id):
                await self.member_repo.add_member(channel["id"], crew_id)

        # Ensure the requesting user is also a member
        if not await self.member_repo.is_member(channel["id"], user_id):
            await self.member_repo.add_member(channel["id"], user_id)

        return channel

    async def create_dm_channel(
        self, user_ids: list[int], created_by: int
    ) -> dict:
        """Create or find existing DM channel between users."""
        return await self.channel_repo.create_dm_channel(user_ids, created_by)

    async def get_inbox(self, user_id: int) -> dict:
        """Assemble the full inbox for a user.

        Returns channels with unread counts, last message preview,
        and global unread/mention stats for the badge.
        """
        channels = await self.channel_repo.get_user_channels(user_id)

        # Build channel response list with last message preview
        channel_list = []
        total_unread = 0
        for ch in channels:
            unread = ch.get("unread_count", 0)
            total_unread += unread

            # Build last message preview if available
            last_message = None
            if ch.get("last_msg_id"):
                last_message = {
                    "id": ch["last_msg_id"],
                    "content": ch.get("last_msg_content"),
                    "message_type": ch.get("last_msg_type", "text"),
                    "sender_name": ch.get("last_msg_sender"),
                    "created_at": ch.get("last_msg_at"),
                }

            channel_list.append({
                "id": ch["id"],
                "channel_type": ch["channel_type"],
                "job_id": ch.get("job_id"),
                "name": ch.get("name"),
                "created_by": ch.get("created_by"),
                "created_at": ch.get("created_at"),
                "updated_at": ch.get("updated_at"),
                "member_count": ch.get("member_count", 0),
                "unread_count": unread,
                "last_message": last_message,
                "job_name": ch.get("job_name"),
                "job_number": ch.get("job_number"),
            })

        unread_mentions = await self.mention_repo.count_unread(user_id)

        return {
            "channels": channel_list,
            "total_unread": total_unread,
            "unread_mentions": unread_mentions,
        }

    async def get_channel_detail(
        self,
        channel_id: int,
        user_id: int,
        *,
        before_id: int | None = None,
        limit: int = 50,
    ) -> dict:
        """Get channel info + paginated messages + members + pinned.

        Also marks the channel as read up to the latest message.
        """
        # Verify membership
        if not await self.member_repo.is_member(channel_id, user_id):
            raise PermissionError("Not a member of this channel")

        # Fetch data in parallel-friendly order
        channel = dict(await self.channel_repo.get_by_id(channel_id))
        messages = await self.message_repo.get_messages(
            channel_id, before_id=before_id, limit=limit + 1
        )

        has_more = len(messages) > limit
        if has_more:
            messages = messages[1:]  # Remove oldest (we fetched limit+1 DESC)

        members = await self.member_repo.get_channel_members(channel_id)
        pinned = await self.message_repo.get_pinned(channel_id)

        # Auto-mark as read (latest message in this batch)
        if messages:
            latest_id = messages[-1]["id"]
            await self.read_repo.mark_read(channel_id, user_id, latest_id)

        return {
            "channel": channel,
            "messages": messages,
            "members": members,
            "pinned_messages": pinned,
            "has_more": has_more,
        }

    async def get_channel_members(self, channel_id: int) -> list[dict]:
        """Get all members of a channel."""
        return await self.member_repo.get_channel_members(channel_id)

    async def add_member(
        self, channel_id: int, user_id: int, role: str = "member"
    ) -> int:
        """Add a member to a channel."""
        return await self.member_repo.add_member(channel_id, user_id, role)

    async def remove_member(self, channel_id: int, user_id: int) -> bool:
        """Remove a member from a channel."""
        return await self.member_repo.remove_member(channel_id, user_id)

    # ── Messages ───────────────────────────────────────────────────

    async def send_message(
        self,
        channel_id: int,
        sender_id: int,
        content: str | None,
        *,
        message_type: str = "text",
        media_path: str | None = None,
        media_mime_type: str | None = None,
        media_size_bytes: int | None = None,
        reply_to_id: int | None = None,
        mention_ids: list[int] | None = None,
        qa_thread_id: int | None = None,
        qa_level: str | None = None,
    ) -> dict:
        """Send a message with membership validation and @mention processing.

        Flow:
        1. Validate sender is channel member
        2. Create message record
        3. Create mention records for each mentioned user
        4. Fire notifications for channel members + mentioned users
        5. Return message with sender info
        """
        # 1. Validate membership
        if not await self.member_repo.is_member(channel_id, sender_id):
            raise PermissionError("Not a member of this channel")

        # 2. Create message
        msg = await self.message_repo.create_message(
            channel_id=channel_id,
            sender_id=sender_id,
            content=content,
            message_type=message_type,
            media_path=media_path,
            media_mime_type=media_mime_type,
            media_size_bytes=media_size_bytes,
            reply_to_id=reply_to_id,
            qa_thread_id=qa_thread_id,
            qa_level=qa_level,
        )

        # 3. Process @mentions
        if mention_ids:
            # Filter to only channel members
            valid_mentions = []
            for mid in mention_ids:
                if mid != sender_id and await self.member_repo.is_member(
                    channel_id, mid
                ):
                    valid_mentions.append(mid)

            if valid_mentions:
                await self.mention_repo.create_mentions(
                    msg["id"], valid_mentions
                )

                # Fire mention notifications
                await self.notif_service.notify(
                    "chat_mention",
                    f"@mentioned you in chat",
                    message=(content or "")[:100],
                    link=f"/chat/inbox?channel={channel_id}",
                    entity_type="chat_message",
                    entity_id=msg["id"],
                    target_user_ids=valid_mentions,
                )

        # 4. Fire message notification to other members (not sender, not mentioned)
        members = await self.member_repo.get_channel_members(channel_id)
        other_ids = [
            m["user_id"]
            for m in members
            if m["user_id"] != sender_id
            and m["user_id"] not in (mention_ids or [])
        ]
        if other_ids:
            sender_name = msg.get("sender_name", "Someone")
            preview = (content or "[media]")[:60]
            await self.notif_service.notify(
                "chat_message",
                f"{sender_name}: {preview}",
                link=f"/chat/inbox?channel={channel_id}",
                entity_type="chat_message",
                entity_id=msg["id"],
                target_user_ids=other_ids,
            )

        return msg

    async def edit_message(
        self, message_id: int, content: str, editor_id: int
    ) -> bool:
        """Edit a message. Only the sender can edit."""
        row = await self.message_repo.get_by_id(message_id)
        if not row:
            raise ValueError("Message not found")
        if row["sender_id"] != editor_id:
            raise PermissionError("Can only edit your own messages")
        return await self.message_repo.edit_content(message_id, content)

    async def delete_message(self, message_id: int, deleter_id: int) -> bool:
        """Soft-delete a message. Sender or channel admin can delete."""
        row = await self.message_repo.get_by_id(message_id)
        if not row:
            raise ValueError("Message not found")
        if row["sender_id"] != deleter_id:
            raise PermissionError("Can only delete your own messages")
        return await self.message_repo.soft_delete(message_id)

    async def pin_message(self, message_id: int, pinned_by: int) -> bool:
        """Pin a message in its channel."""
        return await self.message_repo.pin(message_id, pinned_by)

    async def unpin_message(self, message_id: int) -> bool:
        """Unpin a message."""
        return await self.message_repo.unpin(message_id)

    # ── Read Receipts ──────────────────────────────────────────────

    async def mark_read(
        self, channel_id: int, user_id: int, message_id: int
    ) -> None:
        """Mark a channel as read up to a specific message."""
        await self.read_repo.mark_read(channel_id, user_id, message_id)

    # ── Mentions ───────────────────────────────────────────────────

    async def get_unread_mentions(self, user_id: int) -> list[dict]:
        """Get all unacknowledged @mentions for a user."""
        return await self.mention_repo.get_unread_mentions(user_id)

    async def acknowledge_mention(self, mention_id: int) -> bool:
        """Mark a mention as acknowledged."""
        return await self.mention_repo.acknowledge(mention_id)

    # ── Internal Helpers ───────────────────────────────────────────

    async def _get_job_crew(self, job_id: int) -> list[int]:
        """Get all user IDs associated with a job.

        Combines:
        - job_team_members (permanent crew assignments)
        - job_dispatch (daily dispatched workers)
        - jobs.lead_user_id (job lead)
        """
        user_ids: set[int] = set()

        # Job lead
        cursor = await self.db.execute(
            "SELECT lead_user_id FROM jobs WHERE id = ?", (job_id,)
        )
        job = await cursor.fetchone()
        if job and job["lead_user_id"]:
            user_ids.add(job["lead_user_id"])

        # Team members (if table exists — added in migration 037)
        try:
            cursor = await self.db.execute(
                "SELECT user_id FROM job_team_members WHERE job_id = ?",
                (job_id,),
            )
            for row in await cursor.fetchall():
                user_ids.add(row["user_id"])
        except Exception:
            pass  # Table may not exist yet

        # Dispatched workers (last 30 days of dispatches)
        try:
            cursor = await self.db.execute(
                """SELECT DISTINCT user_id FROM job_dispatch
                   WHERE job_id = ? AND dispatch_date >= date('now', '-30 days')""",
                (job_id,),
            )
            for row in await cursor.fetchall():
                user_ids.add(row["user_id"])
        except Exception:
            pass

        return list(user_ids)

    async def get_badge_count(self, user_id: int) -> dict:
        """Get combined unread count for the chat badge in the nav."""
        channels = await self.channel_repo.get_user_channels(user_id)
        total_unread = sum(ch.get("unread_count", 0) for ch in channels)
        unread_mentions = await self.mention_repo.count_unread(user_id)
        return {
            "total_unread": total_unread,
            "unread_mentions": unread_mentions,
        }

File: frontend/src/local/services/chat-service.ts
/**
 * Local Chat Service — offline read/write for messages, channels, and Q&A.
 *
 * Mirrors backend/app/services/chat_service.py + qa_service.py
 * for field workers. Supports: channel list/detail, message CRUD,
 * read receipts, Q&A threads, and offline message queuing.
 *
 * RFI management is shop-only — synced down for visibility only.
 */

import { getDb } from '../db';
import { trackChange } from '../change-tracker';
import { BaseRepo } from '../repos/base-repo';

// ── Repos ──────────────────────────────────────────────────────────

const channelRepo = new BaseRepo('chat_channels');
const memberRepo = new BaseRepo('chat_channel_members');
const messageRepo = new BaseRepo('chat_messages');
const mentionRepo = new BaseRepo('chat_mentions');
const qaRepo = new BaseRepo('qa_threads');

// ── Types ──────────────────────────────────────────────────────────

export interface LocalChannel {
  id: number;
  channel_type: string;
  job_id: number | null;
  name: string | null;
  created_by: number;
  is_active: number;
  created_at: string;
  updated_at: string;
  // Joined
  job_name?: string;
  job_number?: string;
  display_name?: string;
  unread_count?: number;
  last_message_preview?: string;
  last_message_at?: string;
  member_count?: number;
}

export interface LocalMessage {
  id: number;
  channel_id: number;
  sender_id: number;
  message_type: string;
  content: string | null;
  media_path: string | null;
  reply_to_id: number | null;
  pinned_at: string | null;
  qa_thread_id: number | null;
  qa_level: string | null;
  edited_at: string | null;
  deleted_at: string | null;
  created_at: string;
  // Joined
  sender_name?: string;
  reply_preview?: string;
  reply_sender_name?: string;
}

export interface LocalQAThread {
  id: number;
  channel_id: number | null;
  job_id: number;
  asked_by: number;
  subject: string;
  current_level: string;
  assigned_to: number | null;
  status: string;
  priority: string;
  answer_text: string | null;
  answered_by: number | null;
  answered_at: string | null;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
  // Joined
  asker_name?: string;
  assigned_name?: string;
  job_number?: string;
}

// ── Inbox ──────────────────────────────────────────────────────────

/** Get channels the user belongs to, with unread counts and last message preview */
export async function getInbox(userId: number): Promise<LocalChannel[]> {
  const db = await getDb();

  const result = await db.query(
    `SELECT cc.*,
       j.job_name, j.job_number,
       (SELECT COUNT(*) FROM chat_channel_members WHERE channel_id = cc.id) as member_count,
       (SELECT cm.content FROM chat_messages cm
        WHERE cm.channel_id = cc.id AND cm.deleted_at IS NULL
        ORDER BY cm.created_at DESC LIMIT 1) as last_message_preview,
       (SELECT cm.created_at FROM chat_messages cm
        WHERE cm.channel_id = cc.id AND cm.deleted_at IS NULL
        ORDER BY cm.created_at DESC LIMIT 1) as last_message_at,
       (SELECT COUNT(*) FROM chat_messages cm
        WHERE cm.channel_id = cc.id
          AND cm.deleted_at IS NULL
          AND cm.sender_id != ?
          AND cm.id > COALESCE(
            (SELECT last_read_message_id FROM chat_read_receipts
             WHERE channel_id = cc.id AND user_id = ?), 0)
       ) as unread_count
     FROM chat_channels cc
     JOIN chat_channel_members ccm ON ccm.channel_id = cc.id
     LEFT JOIN jobs j ON j.id = cc.job_id
     WHERE ccm.user_id = ? AND ccm.left_at IS NULL AND cc.is_active = 1
     ORDER BY last_message_at DESC NULLS LAST`,
    [userId, userId, userId],
  );
  return result.values as LocalChannel[];
}

// ── Channel Detail ────────────────────────────────────────────────

/** Get messages for a channel (cursor-based, newest first) */
export async function getChannelMessages(
  channelId: number,
  opts?: { before_id?: number; limit?: number },
): Promise<{ messages: LocalMessage[]; has_more: boolean }> {
  const db = await getDb();
  const limit = opts?.limit ?? 50;

  const conditions = ['cm.channel_id = ?', 'cm.deleted_at IS NULL'];
  const params: any[] = [channelId];

  if (opts?.before_id) {
    conditions.push('cm.id < ?');
    params.push(opts.before_id);
  }

  const where = conditions.join(' AND ');
  params.push(limit + 1); // +1 for has_more

  const result = await db.query(
    `SELECT cm.*,
       u.display_name as sender_name,
       reply.content as reply_preview,
       ru.display_name as reply_sender_name
     FROM chat_messages cm
     JOIN users u ON u.id = cm.sender_id
     LEFT JOIN chat_messages reply ON reply.id = cm.reply_to_id
     LEFT JOIN users ru ON ru.id = reply.sender_id
     WHERE ${where}
     ORDER BY cm.created_at DESC
     LIMIT ?`,
    params,
  );

  const rows = result.values as LocalMessage[];
  const has_more = rows.length > limit;
  if (has_more) rows.pop();

  // Return in chronological order (oldest first)
  return { messages: rows.reverse(), has_more };
}

/** Get channel members */
export async function getChannelMembers(channelId: number) {
  const db = await getDb();
  const result = await db.query(
    `SELECT ccm.*, u.display_name, u.username
     FROM chat_channel_members ccm
     JOIN users u ON u.id = ccm.user_id
     WHERE ccm.channel_id = ? AND ccm.left_at IS NULL`,
    [channelId],
  );
  return result.values;
}

/** Get pinned messages for a channel */
export async function getPinnedMessages(channelId: number): Promise<LocalMessage[]> {
  const db = await getDb();
  const result = await db.query(
    `SELECT cm.*, u.display_name as sender_name
     FROM chat_messages cm
     JOIN users u ON u.id = cm.sender_id
     WHERE cm.channel_id = ? AND cm.pinned_at IS NOT NULL AND cm.deleted_at IS NULL
     ORDER BY cm.pinned_at DESC`,
    [channelId],
  );
  return result.values as LocalMessage[];
}

// ── Message CRUD ──────────────────────────────────────────────────

/** Send a message (queued for sync) */
export async function sendMessage(
  channelId: number,
  senderId: number,
  opts: {
    content?: string;
    message_type?: string;
    media_path?: string;
    reply_to_id?: number;
    mention_ids?: number[];
  },
): Promise<number> {
  const id = await messageRepo.insert({
    channel_id: channelId,
    sender_id: senderId,
    message_type: opts.message_type ?? 'text',
    content: opts.content ?? null,
    media_path: opts.media_path ?? null,
    reply_to_id: opts.reply_to_id ?? null,
  });

  await trackChange('chat_messages', id, 'INSERT');

  // Create mention records
  if (opts.mention_ids?.length) {
    for (const mentionedUserId of opts.mention_ids) {
      const mId = await mentionRepo.insert({
        message_id: id,
        mentioned_user_id: mentionedUserId,
      });
      await trackChange('chat_mentions', mId, 'INSERT');
    }
  }

  // Update channel's updated_at
  await channelRepo.update(channelId, { updated_at: new Date().toISOString() });
  await trackChange('chat_channels', channelId, 'UPDATE');

  return id;
}

/** Edit a message */
export async function editMessage(messageId: number, content: string): Promise<void> {
  await messageRepo.update(messageId, {
    content,
    edited_at: new Date().toISOString(),
  });
  await trackChange('chat_messages', messageId, 'UPDATE');
}

/** Soft-delete a message */
export async function deleteMessage(messageId: number): Promise<void> {
  await messageRepo.update(messageId, {
    deleted_at: new Date().toISOString(),
  });
  await trackChange('chat_messages', messageId, 'UPDATE');
}

/** Pin/unpin a message */
export async function pinMessage(messageId: number, userId: number): Promise<void> {
  await messageRepo.update(messageId, {
    pinned_at: new Date().toISOString(),
    pinned_by: userId,
  });
  await trackChange('chat_messages', messageId, 'UPDATE');
}

export async function unpinMessage(messageId: number): Promise<void> {
  await messageRepo.update(messageId, {
    pinned_at: null,
    pinned_by: null,
  });
  await trackChange('chat_messages', messageId, 'UPDATE');
}

// ── Read Receipts ─────────────────────────────────────────────────

/** Mark channel as read up to a message ID (UPSERT) */
export async function markChannelRead(
  channelId: number,
  userId: number,
  lastReadMessageId: number,
): Promise<void> {
  const db = await getDb();
  await db.run(
    `INSERT INTO chat_read_receipts (channel_id, user_id, last_read_message_id, read_at)
     VALUES (?, ?, ?, datetime('now'))
     ON CONFLICT(channel_id, user_id)
     DO UPDATE SET last_read_message_id = MAX(last_read_message_id, excluded.last_read_message_id),
                   read_at = datetime('now')`,
    [channelId, userId, lastReadMessageId],
  );
}

// ── Q&A Threads ───────────────────────────────────────────────────

/** List Q&A threads with optional filters */
export async function listQAThreads(opts?: {
  job_id?: number;
  status?: string;
  assigned_to?: number;
  limit?: number;
}): Promise<LocalQAThread[]> {
  const db = await getDb();
  const conditions: string[] = [];
  const params: any[] = [];

  if (opts?.job_id) {
    conditions.push('qt.job_id = ?');
    params.push(opts.job_id);
  }
  if (opts?.status) {
    conditions.push('qt.status = ?');
    params.push(opts.status);
  }
  if (opts?.assigned_to) {
    conditions.push('qt.assigned_to = ?');
    params.push(opts.assigned_to);
  }

  const where = conditions.length ? `WHERE ${conditions.join(' AND ')}` : '';
  params.push(opts?.limit ?? 100);

  const result = await db.query(
    `SELECT qt.*,
       ua.display_name as asker_name,
       uas.display_name as assigned_name,
       j.job_number
     FROM qa_threads qt
     JOIN users ua ON ua.id = qt.asked_by
     LEFT JOIN users uas ON uas.id = qt.assigned_to
     JOIN jobs j ON j.id = qt.job_id
     ${where}
     ORDER BY qt.updated_at DESC
     LIMIT ?`,
    params,
  );
  return result.values as LocalQAThread[];
}

/** Ask a new question — creates a Q&A thread + system message */
export async function askQuestion(
  jobId: number,
  askedBy: number,
  subject: string,
  _body: string,
  priority: string = 'normal',
): Promise<number> {
  const id = await qaRepo.insert({
    job_id: jobId,
    asked_by: askedBy,
    subject,
    status: 'open',
    priority,
    current_level: 'worker',
  });
  await trackChange('qa_threads', id, 'INSERT');
  return id;
}

/** Answer a Q&A thread */
export async function answerQuestion(
  threadId: number,
  answeredBy: number,
  answerText: string,
): Promise<void> {
  await qaRepo.update(threadId, {
    status: 'answered',
    answer_text: answerText,
    answered_by: answeredBy,
    answered_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
  });
  await trackChange('qa_threads', threadId, 'UPDATE');
}

// ── Channel Creation ──────────────────────────────────────────────

/** Find or create a job channel */
export async function getOrCreateJobChannel(
  jobId: number,
  createdBy: number,
): Promise<number> {
  const db = await getDb();

  // Check for existing
  const existing = await db.query(
    `SELECT id FROM chat_channels WHERE channel_type = 'job' AND job_id = ? AND is_active = 1`,
    [jobId],
  );

  if (existing.values.length > 0) {
    return existing.values[0].id;
  }

  // Create new channel
  const id = await channelRepo.insert({
    channel_type: 'job',
    job_id: jobId,
    created_by: createdBy,
  });
  await trackChange('chat_channels', id, 'INSERT');

  // Add creator as first member
  const mId = await memberRepo.insert({
    channel_id: id,
    user_id: createdBy,
    role: 'admin',
  });
  await trackChange('chat_channel_members', mId, 'INSERT');

  return id;
}
